grpcadapter/adapter.go

```diff
@@ -0,0 +1,770 @@
+
+		isCompressed := false
+		if contentEncoding := query.Get(connectUnaryCompressionQueryParameter); len(contentEncoding) > 0 {
+			isCompressed = contentEncoding != "identity"
+			delHeaderCanonical(header, connectUnaryHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+
+		msg := query.Get(connectUnaryMessageQueryParameter)
+		msgReader := queryValueReader(msg, query.Get(connectUnaryBase64QueryParameter) == "1")
+		request.Body = &bufferedEnvelopeReader{
+			ReadCloser:   io.NopCloser(msgReader),
+			isCompressed: isCompressed,
+			maxBytes:     readMaxBytes,
+		}
+	} else {
+		isCompressed := false
+		if contentEncoding := getHeaderCanonical(header, connectUnaryHeaderCompression); len(contentEncoding) > 0 {
+			isCompressed = contentEncoding != "identity"
+			delHeaderCanonical(header, connectUnaryHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+		request.Body = &bufferedEnvelopeReader{
+			ReadCloser:   request.Body,
+			isCompressed: isCompressed,
+			maxBytes:     readMaxBytes,
+		}
+	}
+}
+
+func isProtocolHeader(header string) bool {
+	return header == connectUnaryHeaderCompression ||
+		header == connectUnaryHeaderAcceptCompression ||
+		header == connectStreamingHeaderCompression ||
+		header == connectStreamingHeaderAcceptCompression ||
+		header == connectHeaderTimeout ||
+		header == connectHeaderProtocolVersion ||
+		header == grpcHeaderCompression ||
+		header == grpcHeaderAcceptCompression ||
+		header == grpcHeaderTimeout ||
+		header == grpcHeaderStatus ||
+		header == grpcHeaderMessage ||
+		header == grpcHeaderDetails
+}
+
+type connectStreamingResponseWriter struct {
+	http.ResponseWriter
+	contentType string
+	config      *config
+	header      http.Header
+	wroteHeader bool
+}
+
+func newConnectStreamingResponseWriter(responseWriter http.ResponseWriter, contentType string, config *config) *connectStreamingResponseWriter {
+	return &connectStreamingResponseWriter{
+		ResponseWriter: responseWriter,
+		contentType:    contentType,
+		config:         config,
+	}
+}
+
+func (w *connectStreamingResponseWriter) Header() http.Header {
+	if w.header == nil {
+		w.header = make(http.Header)
+	}
+	return w.header
+}
+
+func (w *connectStreamingResponseWriter) writeHeader(statusCode int) {
+	// Encode header response.
+	header := w.ResponseWriter.Header()
+	compression := getHeaderCanonical(w.header, grpcHeaderCompression)
+
+	setHeaderCanonical(header, headerContentType, w.contentType)
+	setHeaderCanonical(header, connectStreamingHeaderCompression, compression)
+
+	for rawKey, values := range w.header {
+		key := textproto.CanonicalMIMEHeaderKey(rawKey)
+		isTrailer := strings.HasPrefix(key, http.TrailerPrefix)
+		if isProtocolHeader(key) || isTrailer {
+			continue
+		}
+		if rawKey != key {
+			delete(header, rawKey)
+		}
+		header[key] = values
+	}
+	if statusCode != http.StatusOK {
+		w.ResponseWriter.WriteHeader(statusCode)
+	}
```

Comment:
Why not call the underlying ResponseWriter's WriteHeader method for 200s?

Reply:
Fine for streaming, not for unary. Fixed.